File: Card.py
from typing import Iterable, Union


class Card:
    SUITS = ["♥", "♦", "♣", "♠"]
    VALUES = ["2", "3", "4", "5", "6", "7", "8", "9", "10", "J", "Q", "K", "A"]

    def __init__(self, rank: Union[int, str], suit: Union[int, str]):
        if isinstance(rank, str):
            rank = Card.VALUES.index(rank)
        if isinstance(suit, str):
            suit = Card.SUITS.index(suit)
        self.rank = rank
        self.suit = suit

    def __str__(self):
        return Card.SUITS[self.suit] + " " + Card.VALUES[self.rank]

    def __repr__(self) -> str:
        return f"Card({self.rank}, {self.suit})"

    def get_cli_repr(self):
        return """
┌───────┐
│{}     │
│       │
│   {}  │
│       │
│     {}│
└───────┘""".format(
            Card.VALUES[self.rank].ljust(2),
            Card.SUITS[self.suit].ljust(2),
            Card.VALUES[self.rank].rjust(2),
        )

    @classmethod
    def from_index(cls, i):
        suit = i // 13
        rank = i % 13
        return cls(rank, suit)

    def to_index(self):
        return self.suit * 13 + self.rank

    @classmethod
    def get_rank(cls, value: str):
        return cls.VALUES.index(value)

    @classmethod
    def get_cli_repr_for_cards(cls, cards: Iterable[Union[int, "Card"]]):
        return "\n".join(
            [
                " ".join(parts)
                for parts in zip(
                    *[
                        (c if isinstance(c, Card) else Card.from_index(c))
                        .get_cli_repr()
                        .split("\n")
                        for c in cards
                    ]
                )
            ]
        )

    def __hash__(self) -> int:
        return self.to_index()

    def __gt__(self, other):
        return self.rank > other.rank

    def __lt__(self, other):
        return self.rank < other.rank

    def __eq__(self, value: object) -> bool:
        if not isinstance(value, Card):
            return False
        return self.rank == value.rank and self.suit == value.suit


if __name__ == "__main__":
    # Example: card 0 is the 2 of hearts
    print(Card.from_index(0))
    print(Card.from_index(0).get_cli_repr())
    # Example 2: card 51 is the ace of spades
    print(Card.from_index(51))
    print(Card.from_index(51).get_cli_repr())

File: Card.cpp
#include "Card.hpp"
#include <sstream>
#include <iomanip>
#include <algorithm>

const std::vector<std::string> Card::SUITS = {"♥", "♦", "♣", "♠"};
const std::vector<std::string> Card::VALUES = {"2", "3", "4", "5", "6", "7", "8", "9", "10", "J", "Q", "K", "A"};

Card::Card() : rank(0), suit(0) {}

Card::Card(int rank, int suit) : rank(rank), suit(suit) {}

Card::Card(std::string rank, std::string suit)
{
    this->rank = std::distance(VALUES.begin(), std::find(VALUES.begin(), VALUES.end(), rank));
    this->suit = std::distance(SUITS.begin(), std::find(SUITS.begin(), SUITS.end(), suit));
}

Card::Card(int index)
{
    this->rank = index % 13;
    this->suit = index / 13;
}

int Card::to_index() const
{
    return suit * 13 + rank;
}

std::string Card::term_color() const
{
    return (suit == 0 || suit == 1) ? TerminalColors::RED : TerminalColors::BLUE;
}

std::string Card::get_cli_repr() const
{
    std::stringstream ss;
    ss << "┌───────┐\n";
    ss << "│" << term_color() << std::left << std::setw(2) << VALUES[rank] << TerminalColors::DEFAULT << "     │\n";
    ss << "│       │\n";
    ss << "│   " << term_color() << SUITS[suit] << TerminalColors::DEFAULT << "   │\n";
    ss << "│       │\n";
    ss << "│     " << term_color() << std::right << std::setw(2) << VALUES[rank] << TerminalColors::DEFAULT << "│\n";
    ss << "└───────┘";
    return ss.str();
}

std::string Card::get_cli_repr_for_cards(const std::vector<Card> &cards)
{
    std::vector<std::vector<std::string>> lines(cards.size(), std::vector<std::string>(7));
    for (size_t i = 0; i < cards.size(); ++i)
    {
        std::stringstream ss(cards[i].get_cli_repr());
        std::string line;
        size_t line_no = 0;
        while (std::getline(ss, line))
        {
            lines[i][line_no++] = line;
        }
    }

    std::string combined;
    for (size_t i = 0; i < 7; ++i)
    { // 7 lines per card
        for (const auto &line_vec : lines)
        {
            combined += line_vec[i] + " ";
        }
        combined += "\n";
    }
    return combined;
}

std::string Card::get_cli_repr_for_cards(const std::vector<int> &cards)
{
    std::vector<Card> card_objs;
    for (int card : cards)
    {
        card_objs.push_back(Card(card));
    }
    return get_cli_repr_for_cards(card_objs);
}

int Card::get_rank(const std::string &rank)
{
    return std::distance(VALUES.begin(), std::find(VALUES.begin(), VALUES.end(), rank));
}

std::string Card::str() const
{
    return SUITS[suit] + " " + VALUES[rank];
}

bool Card::operator<(const Card &other) const
{
    if (rank != other.rank)
        return rank < other.rank;
    return suit < other.suit;
}

bool Card::operator>(const Card &other) const
{
    if (rank != other.rank)
        return rank > other.rank;
    return suit > other.suit;
}

bool Card::operator==(const Card &other) const
{
    return rank == other.rank && suit == other.suit;
}
